tests/test_moisture_sensor.py

```diff
@@ -7,14 +7,21 @@
 
 class MoistureSensorTest(unittest.TestCase):
 
-    def setUp(self):
-        self.mock_adc = mock.Mock()
+    def test_moisture_50(self):
+        """Moisture level read should be the average of the two readings."""
+        mock_adc = mock.Mock()
+        mock_adc.read_adc.side_effect = [300, (1023 - 400)]
+        mock_pi_io = mock.Mock()
+        mock_clock = mock.Mock()
         channel = 1
-        self.moisture_sensor = moisture_sensor.MoistureSensor(self.mock_adc,
-                                                              channel)
+        gpio_pins = (12, 16)
+        sensor = moisture_sensor.MoistureSensor(mock_adc, mock_pi_io, channel,
+                                                gpio_pins, mock_clock)
+        moisture_level = sensor.moisture()
 
-    def test_moisture_50(self):
-        """Moisture level read should be moisture level returned."""
-        self.mock_adc.read_adc.return_value = 50
-        moisture_level = self.moisture_sensor.moisture()
-        self.assertEqual(moisture_level, 50)
+        self.assertIn(mock.call(12), mock_pi_io.turn_pin_on.call_args_list)
+        self.assertIn(mock.call(12), mock_pi_io.turn_pin_off.call_args_list)
+        self.assertIn(mock.call(16), mock_pi_io.turn_pin_on.call_args_list)
+        self.assertIn(mock.call(16), mock_pi_io.turn_pin_off.call_args_list)
```

Comment:
The problem here is that 22 and 23 are redundant, as are 24 and 25 (it would still pass with only one call). And it hides the fact that we're actually calling `turn_pin_off` three times.

Suggest checking like this:

```python
        self.assertEqual([mock.call(12), mock.call(16)],
                         mock_pi_io.turn_pin_on.call_args_list)
        self.assertEqual([mock.call(16), mock.call(12), mock.call(16)],
                         mock_pi_io.turn_pin_off.call_args_list)
```